decoder/label_candidates.go

```diff
@@ -35,44 +37,56 @@ func (d *Decoder) labelCandidatesFromDependentSchema(idx int, db map[schema.Sche
 		bodySchema := db[schemaKey]
 
 		for _, label := range depKeys.Labels {
-			if label.Index == idx {
-				if len(prefix) > 0 && !strings.HasPrefix(label.Value, string(prefix)) {
-					continue
-				}
+			if label.Index != idx {
+				continue
+			}
 
-				// Dependent keys may be duplicated where one
-				// key is labels-only and other one contains
-				// labels + attributes.
-				//
-				// Specifically in Terraform this applies to
-				// a resource type depending on 'provider' attribute.
-				//
-				// We do need such dependent keys elsewhere
-				// to know how to do completion within a block
-				// but this doesn't matter when completing the label itself
-				// unless/until we're also completing the dependent attributes.
-				if _, ok := foundCandidateNames[label.Value]; ok {
-					continue
-				}
+			if len(prefix) > 0 && !strings.HasPrefix(label.Value, string(prefix)) {
+				continue
+			}
 
-				candidates.List = append(candidates.List, lang.Candidate{
-					Label:        label.Value,
-					Kind:         lang.LabelCandidateKind,
-					IsDeprecated: bodySchema.IsDeprecated,
-					TextEdit: lang.TextEdit{
-						NewText: label.Value,
-						Snippet: label.Value,
-						Range:   editRng,
-					},
-					// TODO: AdditionalTextEdits:
-					// - prefill required fields if body is empty
-					// - prefill dependent attribute(s)
-					Detail:      bodySchema.Detail,
-					Description: bodySchema.Description,
-				})
-				foundCandidateNames[label.Value] = true
-				count++
+			// Dependent keys may be duplicated where one
+			// key is labels-only and other one contains
+			// labels + attributes.
+			//
+			// Specifically in Terraform this applies to
+			// a resource type depending on 'provider' attribute.
+			//
+			// We do need such dependent keys elsewhere
+			// to know how to do completion within a block
+			// but this doesn't matter when completing the label itself
+			// unless/until we're also completing the dependent attributes.
+			if _, ok := foundCandidateNames[label.Value]; ok {
+				continue
+			}
+
+			te := lang.TextEdit{}
+			if d.PrefillRequiredAttributes {
+				snippet := generateSnippet(label.Value, bodySchema.Attributes)
```

Comment:
As mentioned in Slack we'll also need to reflect "required" blocks here, i.e. range over `bodySchema.Blocks` and add any blocks with `MinItems > 0`.

One related question which may also come up is whether we also want to generate required attributes _within_ these blocks. Personally, I think that even empty blocks would be useful as a first iteration (i.e. I wouldn't worry about it _too much_) - but if you can get it all working then great!

btw. The story of blocks in provider SDK is historically tricky and they will likely be phased out (see https://github.com/hashicorp/terraform-plugin-framework/issues/85), but we should still support them, also more importantly `hcl-lang` doesn't focus only on provider schema, nor just Terraform.